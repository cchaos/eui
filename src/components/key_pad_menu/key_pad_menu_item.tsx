/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  FunctionComponent,
  ReactNode,
  HTMLAttributes,
  Ref,
} from 'react';
import classNames from 'classnames';

import {
  CommonProps,
  ExclusiveUnion,
  PropsForAnchor,
  PropsForButton,
} from '../common';

import { EuiBetaBadge } from '../badge/beta_badge';

import { getSecureRelForTarget } from '../../services';

import { IconType } from '../icon';

const renderContent = (
  children: ReactNode,
  label: ReactNode,
  betaBadgeLabel?: ReactNode,
  betaBadgeTooltipContent?: ReactNode,
  betaBadgeIconType?: IconType
) => (
  <span className="euiKeyPadMenuItem__inner">
    {betaBadgeLabel && (
      <span className="euiKeyPadMenuItem__betaBadgeWrapper">
        <EuiBetaBadge
          className="euiKeyPadMenuItem__betaBadge"
          label={betaBadgeLabel}
          iconType={betaBadgeIconType}
          tooltipContent={betaBadgeTooltipContent}
        />
      </span>
    )}

    <span className="euiKeyPadMenuItem__icon">{children}</span>

    <p className="euiKeyPadMenuItem__label">{label}</p>
  </span>
);

interface EuiKeyPadMenuItemCommonProps {
  /**
   * Pass an EuiIcon, preferrably `size="l"`
   */
  children: ReactNode;
  isDisabled?: boolean;
  /**
   * Indicate if an item is the current one.
   * Be sure to use `true` AND `false` when acting as a toggle.
   */
  isSelected?: boolean;
  /**
   * The text to display beneath the icon
   */
  label: ReactNode;
  /**
   * Add a badge to the card to label it as "Beta" or other non-GA state
   */
  betaBadgeLabel?: string;
  /**
   * Supply an icon type if the badge should just be an icon
   */
  betaBadgeIconType?: IconType;
  /**
   * Add a description to the beta badge (will appear in a tooltip)
   */
  betaBadgeTooltipContent?: ReactNode;
  rel?: string;
}

type EuiKeyPadMenuItemPropsForAnchor = PropsForAnchor<
  EuiKeyPadMenuItemCommonProps,
  {
    buttonRef?: Ref<HTMLAnchorElement>;
  }
>;

type EuiKeyPadMenuItemPropsForButton = PropsForButton<
  EuiKeyPadMenuItemCommonProps,
  {
    buttonRef?: Ref<HTMLButtonElement>;
  }
>;

export type EuiKeyPadMenuItemProps = CommonProps &
  ExclusiveUnion<
    EuiKeyPadMenuItemPropsForAnchor,
    EuiKeyPadMenuItemPropsForButton
  >;

export const EuiKeyPadMenuItem: FunctionComponent<EuiKeyPadMenuItemProps> = ({
  isDisabled,
  isSelected,
  label,
  children,
  className,
  betaBadgeLabel,
  betaBadgeTooltipContent,
  betaBadgeIconType,
  href,
  rel,
  target,
  buttonRef,
  ...rest
}) => {
  const classes = classNames(
    'euiKeyPadMenuItem',
    {
      'euiKeyPadMenuItem--hasBetaBadge': betaBadgeLabel,
    },
    className
  );

  const Element = href && !isDisabled ? 'a' : 'button';
  const relObj: {
    disabled?: boolean;
    type?: string;
    href?: string;
    rel?: string;
    target?: string;
    'aria-pressed'?: boolean;
    'aria-current'?: boolean;
  } = {};

  if (href && !isDisabled) {
    relObj.href = href;
    relObj.rel = getSecureRelForTarget({ href, target, rel });
    relObj.target = target;
    relObj['aria-current'] = isSelected ? isSelected : undefined;
  } else {
    relObj.disabled = isDisabled;
    relObj.type = 'button';
    relObj['aria-pressed'] = isSelected;
  }

  return (
    <Element
      className={classes}
      {...(relObj as HTMLAttributes<HTMLElement>)}
      {...(rest as HTMLAttributes<HTMLElement>)}
      // ref={buttonRef} HELP!
    >
      {renderContent(
        children,
        label,
        betaBadgeLabel,
        betaBadgeTooltipContent,
        betaBadgeIconType
      )}
    </Element>
  );
};
