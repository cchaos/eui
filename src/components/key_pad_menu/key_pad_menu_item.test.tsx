/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, shallow } from 'enzyme';
import { requiredProps } from '../../test';

import { EuiKeyPadMenuItem } from './key_pad_menu_item';

describe('EuiKeyPadMenuItem', () => {
  test('is rendered', () => {
    const component = render(
      <EuiKeyPadMenuItem label="Label" {...requiredProps}>
        Icon
      </EuiKeyPadMenuItem>
    );

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    describe('isDisabled', () => {
      test('renders with href', () => {
        const component = render(
          <EuiKeyPadMenuItem isDisabled label="Label" href="#">
            Icon
          </EuiKeyPadMenuItem>
        );

        expect(component).toMatchSnapshot();
      });

      test('renders with onClick', () => {
        const onClickHandler = jest.fn();

        const component = render(
          <EuiKeyPadMenuItem isDisabled label="Label" onClick={onClickHandler}>
            Icon
          </EuiKeyPadMenuItem>
        );

        expect(component).toMatchSnapshot();
      });
    });

    describe('isSelected', () => {
      test('renders with href', () => {
        const component = render(
          <EuiKeyPadMenuItem isSelected label="Label" href="#">
            Icon
          </EuiKeyPadMenuItem>
        );

        expect(component).toMatchSnapshot();
      });

      test('renders with onClick', () => {
        const onClickHandler = jest.fn();

        const component = render(
          <EuiKeyPadMenuItem isSelected label="Label" onClick={onClickHandler}>
            Icon
          </EuiKeyPadMenuItem>
        );

        expect(component).toMatchSnapshot();
      });
    });

    describe('betaBadge', () => {
      test('renders', () => {
        const component = render(
          <EuiKeyPadMenuItem betaBadgeLabel="Beta" label="Label">
            Icon
          </EuiKeyPadMenuItem>
        );

        expect(component).toMatchSnapshot();
      });

      test('renders with betaBadgeIconType', () => {
        const component = render(
          <EuiKeyPadMenuItem
            betaBadgeLabel="Beta"
            betaBadgeIconType="bolt"
            label="Label">
            Icon
          </EuiKeyPadMenuItem>
        );

        expect(component).toMatchSnapshot();
      });

      test('renders with betaBadgeTooltipContent', () => {
        const component = render(
          <EuiKeyPadMenuItem
            betaBadgeLabel="Beta"
            betaBadgeTooltipContent="Content"
            label="Label">
            Icon
          </EuiKeyPadMenuItem>
        );

        expect(component).toMatchSnapshot();
      });
    });
  });

  test('renders href', () => {
    const component = render(
      <EuiKeyPadMenuItem label="Label" href="#">
        Icon
      </EuiKeyPadMenuItem>
    );

    expect(component).toMatchSnapshot();
  });

  test('renders href with rel', () => {
    const component = render(
      <EuiKeyPadMenuItem label="Label" href="#" rel="noreferrer">
        Icon
      </EuiKeyPadMenuItem>
    );

    expect(component).toMatchSnapshot();
  });

  test('renders button', () => {
    const onClickHandler = jest.fn();

    const component = render(
      <EuiKeyPadMenuItem label="Label" onClick={onClickHandler}>
        Icon
      </EuiKeyPadMenuItem>
    );

    expect(component).toMatchSnapshot();
  });

  test("onClick isn't called upon instantiation", () => {
    const onClickHandler = jest.fn();

    shallow(
      <EuiKeyPadMenuItem label="Label" onClick={onClickHandler}>
        Icon
      </EuiKeyPadMenuItem>
    );

    expect(onClickHandler).not.toBeCalled();
  });

  test('onClick is called when the button is clicked', () => {
    const onClickHandler = jest.fn();

    const $button = shallow(
      <EuiKeyPadMenuItem label="Label" onClick={onClickHandler}>
        Icon
      </EuiKeyPadMenuItem>
    );

    $button.simulate('click');

    expect(onClickHandler).toBeCalledTimes(1);
  });

  describe('checkable', () => {
    test('renders as radio', () => {
      const component = render(
        <EuiKeyPadMenuItem onChange={() => {}} checkable="single" label="Label">
          Icon
        </EuiKeyPadMenuItem>
      );

      expect(component).toMatchSnapshot();
    });

    test('renders as checkbox', () => {
      const component = render(
        <EuiKeyPadMenuItem onChange={() => {}} checkable="multi" label="Label">
          Icon
        </EuiKeyPadMenuItem>
      );

      expect(component).toMatchSnapshot();
    });
  });
});
