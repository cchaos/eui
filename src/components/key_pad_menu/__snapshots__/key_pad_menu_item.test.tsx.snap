// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiKeyPadMenuItem is rendered 1`] = `
<button
  aria-label="aria-label"
  class="euiKeyPadMenuItem testClass1 testClass2"
  data-test-subj="test subject string"
  type="button"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</button>
`;

exports[`EuiKeyPadMenuItem props betaBadge renders 1`] = `
<button
  class="euiKeyPadMenuItem euiKeyPadMenuItem--hasBetaBadge"
  type="button"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__betaBadgeWrapper"
    >
      <span
        class="euiBetaBadge euiKeyPadMenuItem__betaBadge"
        title="Beta"
      >
        Beta
      </span>
    </span>
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</button>
`;

exports[`EuiKeyPadMenuItem props betaBadge renders with betaBadgeIconType 1`] = `
<button
  class="euiKeyPadMenuItem euiKeyPadMenuItem--hasBetaBadge"
  type="button"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__betaBadgeWrapper"
    >
      <span
        class="euiBetaBadge euiBetaBadge--iconOnly euiKeyPadMenuItem__betaBadge"
        title="Beta"
      >
        <span
          aria-hidden="true"
          class="euiBetaBadge__icon"
          data-euiicon-type="bolt"
        />
      </span>
    </span>
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</button>
`;

exports[`EuiKeyPadMenuItem props betaBadge renders with betaBadgeTooltipContent 1`] = `
<button
  class="euiKeyPadMenuItem euiKeyPadMenuItem--hasBetaBadge"
  type="button"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__betaBadgeWrapper"
    >
      <span
        class="euiToolTipAnchor"
      >
        <span
          class="euiBetaBadge euiKeyPadMenuItem__betaBadge"
          tabindex="0"
        >
          Beta
        </span>
      </span>
    </span>
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</button>
`;

exports[`EuiKeyPadMenuItem props isDisabled renders with href 1`] = `
<button
  class="euiKeyPadMenuItem"
  disabled=""
  type="button"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</button>
`;

exports[`EuiKeyPadMenuItem props isDisabled renders with onClick 1`] = `
<button
  class="euiKeyPadMenuItem"
  disabled=""
  type="button"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</button>
`;

exports[`EuiKeyPadMenuItem props isSelected renders with href 1`] = `
<a
  aria-current="true"
  class="euiKeyPadMenuItem"
  href="#"
  rel="noreferrer"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</a>
`;

exports[`EuiKeyPadMenuItem props isSelected renders with onClick 1`] = `
<button
  aria-pressed="true"
  class="euiKeyPadMenuItem"
  type="button"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</button>
`;

exports[`EuiKeyPadMenuItem renders button 1`] = `
<button
  class="euiKeyPadMenuItem"
  type="button"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</button>
`;

exports[`EuiKeyPadMenuItem renders href 1`] = `
<a
  class="euiKeyPadMenuItem"
  href="#"
  rel="noreferrer"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</a>
`;

exports[`EuiKeyPadMenuItem renders href with rel 1`] = `
<a
  class="euiKeyPadMenuItem"
  href="#"
  rel="noreferrer"
>
  <span
    class="euiKeyPadMenuItem__inner"
  >
    <span
      class="euiKeyPadMenuItem__icon"
    >
      Icon
    </span>
    <p
      class="euiKeyPadMenuItem__label"
    >
      Label
    </p>
  </span>
</a>
`;
